llvm/lib/Target/AMDGPU/AMDGPUSubtarget.cpp

```diff
@@ -105,6 +105,14 @@ GCNSubtarget::initializeSubtargetDependencies(const Triple &TT,
                                         : AMDGPUSubtarget::SOUTHERN_ISLANDS;
   }
 
+  if (!hasFeature(AMDGPU::FeatureWavefrontSize32) &&
+      !hasFeature(AMDGPU::FeatureWavefrontSize64)) {
+    if (getGeneration() >= AMDGPUSubtarget::GFX10)
+      ToggleFeature(AMDGPU::FeatureWavefrontSize32);
+    else
+      ToggleFeature(AMDGPU::FeatureWavefrontSize64);
```

Comment:
I think we can avoid this by having a separate SupportsWave32 feature implied by FeatureWavefrontSize32

Reply:
We still need to select default somewhere. So say a subtarget has FeatureSupportsWave32, then what?

Reply:
I.e. the line `if (getGeneration() >= AMDGPUSubtarget::GFX10)` can be changed to `if (hasFeature(AMDGPU::FeatureSupportsWave32))`, but that's it.

Also I shall mention, this `if` can be avoided altogether because subtargets without an option to select wave size just have a nailed wave64 feature in their definition. This `if` here is just to catch cases when tests use manual -mattr switching it off, but I can remove it and only set wave32 as a default.

Reply:
Just did that, unconditionally add wave32 if none is set.

Reply:
I mean if you use the implies feature, the feature parsing logic should flip the incompatible case for you instead of manually doing it here

Reply:
I do not see a way to tell it is incompatible in the first place. OK, wave32 implies it supports wave32. How does it turn wave32 automatically? How does it tell that wave64 is incompatible with wave32?

Reply:
IMHO to do what you want features must be:

1) Organized in groups.
2) We need to tell features in group are mutually exclusive.
3) There shall be a way to tell this one is the default.

None of that exists.

Reply:
I mean if there's a SupportsWave32 feature, implied by FeatureWavefrontSize32, if you specify wavefrontsize32 to a wave64-only target, the incompatible feature check in the parsing logic will hit and it will assume you specified an invalid target and unset the target-cpu.